### PR TITLE
Add IronBank in Dune V2 for Ethereum and Optimism

### PR DESCRIPTION
**Pull request by mmibmm**
This is a migration from [V1 Abstractions](https://github.com/duneanalytics/spellbook/blob/main/deprecated-dune-v1-abstractions/ethereum/gnosis_protocol_v2/insert_batches.sql) to V2 Dune Engine.

cc @soispoke as IronBank team reviewer.

*For Dune Engine V2*
I've checked that:
General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if adding a new model, I edited the alter table macro to display new database object (table or view) in UI explorer
* [x] if adding a new materialized table, I edited the optimize table macro

Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by gentrexha** [comment]
>cc @gentrexha as IronBank team reviewer.

I believe you've gotten me mixed up. As I'm currently on the CoW Protocol team.

**Comment by mmibmm** [comment]
> > cc @gentrexha as IronBank team reviewer.
> 
> I believe you've gotten me mixed up. As I'm currently on the CoW Protocol team.

File: macros/alter_table_properties.sql [hunk 1]
```diff
@@ -730,6 +730,24 @@ ALTER VIEW archipelago_ethereum.fees  SET TBLPROPERTIES ('dune.public'='true',
                                                         'dune.data_explorer.contributors'='["0xRob"]');
 {% endset %}
 
+{% set ironbank_ethereum_itokens %}
```
Comment (jeff-dude) [inline macros/alter_table_properties.sql:733]: if you merge main from this repo back into your branch, this file will be deleted (as expected)
we will now have a 'post-hook' property in each model config block, like so:
https://github.com/duneanalytics/spellbook/blob/main/models/uniswap/ethereum/uniswap_v1_ethereum_trades.sql

where the four parameters are fed into this macro:
https://github.com/duneanalytics/spellbook/blob/main/macros/expose_spells.sql

note the macro doesn't need edited, just for reference to see how the code has been enhanced

File: macros/optimize_tables.sql [hunk 2]
```diff
@@ -99,6 +99,14 @@ OPTIMIZE archipelago_ethereum.events;
 OPTIMIZE foundation_ethereum.events;
 {% endset %}
 
+{% set ironbank_ethereum_itokens %}
```
Comment (jeff-dude) [inline macros/optimize_tables.sql:102]: this file isn't necessary as both of your objects are being built as views. this only applies to tables.

File: models/ironbank/ethereum/ironbank_ethereum_itokens.sql [hunk 3]
```diff
@@ -0,0 +1,48 @@
+{{ config(
+    materialized = 'view',
```
Comment (jeff-dude) [inline models/ironbank/ethereum/ironbank_ethereum_itokens.sql:2]: this property isn't necessary as it's already in the dbt_project yaml file you added. it's only necessary if you override that value.

File: models/ironbank/ethereum/ironbank_ethereum_sources.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+version: 2
+
+sources:
+  - name: ironbank_ethereum
+    description: "decoded events and function calls for ironbank on ethereum"
+    tables:
```
Comment (jeff-dude) [inline models/ironbank/ethereum/ironbank_ethereum_sources.yml:6]: i notice you've added a bunch of source decoded tables, however, the models added don't use any of the sources. will they be used later in another PR?

File: models/ironbank/optimism/ironbank_optimism_itokens.sql [hunk 5]
```diff
@@ -0,0 +1,31 @@
+{{ config(
+    materialized = 'view',
```
Comment (jeff-dude) [inline models/ironbank/optimism/ironbank_optimism_itokens.sql:2]: same comment here as ethereum model

File: models/ironbank/optimism/ironbank_optimism_sources.yml [hunk 6]
```diff
@@ -0,0 +1,37 @@
+version: 2
+
+sources:
+  - name: ironbank_optimism
+    description: "decoded events and function calls for ironbank on optimism"
+    freshness:
+      warn_after: { count: 12, period: hour }
+      error_after: { count: 24, period: hour }
+    tables:
```
Comment (jeff-dude) [inline models/ironbank/optimism/ironbank_optimism_sources.yml:9]: same comment here as ethereum sources

**Review by jeff-dude — CHANGES_REQUESTED** [review]
thank you for the PR! added a few comments

File: models/ironbank/ethereum/ironbank_ethereum_sources.yml (hunk 4 above)
Comment (mmibmm) [reply models/ironbank/ethereum/ironbank_ethereum_sources.yml:6]: Yes, some will be used in another PR.

File: macros/alter_table_properties.sql (hunk 1 above)
Comment (mmibmm) [reply macros/alter_table_properties.sql]: Got it. Fixed.

File: macros/optimize_tables.sql (hunk 2 above)
Comment (mmibmm) [reply macros/optimize_tables.sql]: Fixed.

File: models/ironbank/ethereum/ironbank_ethereum_itokens.sql (hunk 3 above)
Comment (mmibmm) [reply models/ironbank/ethereum/ironbank_ethereum_itokens.sql]: Got it. Fixed.

File: models/ironbank/optimism/ironbank_optimism_sources.yml (hunk 6 above)
Comment (mmibmm) [reply models/ironbank/optimism/ironbank_optimism_sources.yml:9]: Same response

**Review by jeff-dude — APPROVED** [review]
you should be all set, thank you for the spell! 

please note that views are deployed once a day, you should see no later than tomorrow.

also, do check that the UI explorer shows your views. i want to be sure the post-hook is working as intended

**Pull request merged**